File: problem94.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""

iteratively:

                            1
                    /               \
                    2               3
                /       \       /       \
                4       5       6       7
                
                
stack=[1]
stack=[3,1,2]



"""
#iterative approach
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack=[]
        res=[]
        
        while True:
            while root:
                stack.append(root)
                root=root.left
            if not stack:
                return res
            root=stack.pop()
            res.append(root.val)
            root=root.right
            
        return res
            
            
            

File: problem94.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        
        IList<int> result = new List<int>();
        TreeNode curr = root;
        
        while(curr != null)
        {
            if(curr.left==null)
            {
                result.Add(curr.val);
                curr=curr.right;
            }
            else
            {
                TreeNode prev = curr.left;
                while(prev.right!=null && prev.right!=curr)
                {
                    prev=prev.right;
                }
                
                if(prev.right==null)
                {
                    prev.right=curr;
                    curr=curr.left;
                }
                else
                {
                    prev.right=null;
                    result.Add(curr.val);
                    curr=curr.right;
                }
            }
        }
        return result;
    }
}

//Morris Inorder Traversal
//time: O(N)
//space: O(1)